src/Microsoft.ML.FastTree/Representation/TreeRegressor.cs

```diff
@@ -0,0 +1,136 @@
+        /// occurs. A non-negative returned value means a node (i.e., not a leaf); for example, 2 means the 3rd node in
+        /// the underlying <see cref="_tree"/>. A negative returned value means a leaf; for example, -1 stands for the
+        /// first leaf in the underlying <see cref="_tree"/>.
+        /// </summary>
+        public ReadOnlySpan<int> LteChild => new ReadOnlySpan<int>(_tree.LteChild, 0, _tree.NumNodes);
+        /// <summary>
+        /// <see cref="GtChild"/>[i] is the i-th node's child index used when the two conditions, (1) and (2),
+        /// described in <see cref="LteChild"/>'s document are not true. Its return value follows the format
+        /// used in <see cref="LteChild"/>.
+        /// </summary>
+        public ReadOnlySpan<int> GtChild => new ReadOnlySpan<int>(_tree.GtChild, 0, _tree.NumNodes);
+        /// <summary>
+        /// <see cref="NumericalSplitFeatureIndexes"/>[i] is the feature index used the splitting function of the
+        /// i-th node. This value is valid only if <see cref="CategoricalSplitFlags"/>[i] is false.
+        /// </summary>
+        public ReadOnlySpan<int> NumericalSplitFeatureIndexes => new ReadOnlySpan<int>(_tree.SplitFeatures, 0, _tree.NumNodes);
+        /// <summary>
+        /// <see cref="NumericalSplitThresholds"/>[i] is the threshold on feature indexed by
+        /// <see cref="NumericalSplitFeatureIndexes"/>[i], where i is the i-th node's index
+        /// (for example, i is 1 for the 2nd node in <see cref="_tree"/>).
+        /// </summary>
+        public ReadOnlySpan<float> NumericalSplitThresholds => new ReadOnlySpan<float>(_tree.RawThresholds, 0, _tree.NumNodes);
+        /// <summary>
+        /// Determine the types of splitting function. If <see cref="CategoricalSplitFlags"/>[i] is true, the i-th
+        /// node's uses categorical splitting function. Otherwise, traditional numerical split is used.
+        /// </summary>
+        public ReadOnlySpan<bool> CategoricalSplitFlags => new ReadOnlySpan<bool>(_tree.CategoricalSplit, 0, _tree.NumNodes);
+        /// <summary>
+        /// <see cref="LeafValues"/>[i] is the learned value at the i-th leaf.
+        /// </summary>
+        public ReadOnlySpan<double> LeafValues => new ReadOnlySpan<double>(_tree.LeafValues, 0, _tree.NumLeaves);
+        /// <summary>
+        /// Return categorical thresholds used at node indexed by nodeIndex. If the considered input feature does NOT
+        /// matche any of values returned by <see cref="GetCategoricalSplitFeaturesAt(int)"/>, we call it a
+        /// less-than-threshold event and therefore <see cref="LteChild"/>[nodeIndex] is the child node that input
+        /// should go next. The returned value is valid only if <see cref="CategoricalSplitFlags"/>[nodeIndex] is true.
+        /// </summary>
+        public ReadOnlySpan<int> GetCategoricalSplitFeaturesAt(int nodeIndex) => new ReadOnlySpan<int>(_tree.CategoricalSplitFeatures[nodeIndex]);
```

Comment:
>public ReadOnlySpan<int> GetCategoricalSplitFeaturesAt(int nodeIndex) => new ReadOnlySpan<int>(_tree.CategoricalSplitFeatures[nodeIndex]); [](start = 8, length = 138)

Can we cache these values?

Reply:
I guess we can't do this for ReadOnlySpan.

---
In reply to: [252021978](https://github.com/dotnet/machinelearning/pull/2243#discussion_r252021978) [](ancestors = 252021978)